PCF-ADR-0005-mcad-api-groups-migration.md

```diff
@@ -0,0 +1,113 @@
+
+## Why
+
+It's conventional for open-source projects to use their domain name as their top-level API group, so DNS is leveraged as a global naming system.
+
+The choice of the current MCAD API groups predates the acquisition of the `codeflare.dev` domain, which explains why they do not follow that convention.
+
+Some of the MCAD APIs, mainly the AppWrapper API, is used by downstream projects, as well as end-users.
+From a client standpoint, the impacts of an API group change imply updating the `apiVersion` field of all the MCAD resources they define, as well as updating the group segment of the REST endpoints, created from the CRDs, that they consume.
+
+A migration plan is needed to provide these clients with a smooth transition to the new API groups.
+
+## Goals
+
+The goal is to inform a _progressive_ migration path, i.e., one that does not require all the downstream projects to handle the impacts in locked steps with MCAD.
+A project like [TorchX](https://github.com/pytorch/torchx), which integrates with MCAD[^1] via the AppWrapper API, has its own release cadence, and it’d be very challenging to have all the components migrated within a single migration cycle, without breaking compatibility.
+
+Stated differently, the goal is to have that migration path started as soon as possible, when the new API group is made available, and have each downstream project, and end-users, walking the migration path at their own pace, while they decide to upgrade their dependency.
+
+[^1]: https://pytorch.org/torchx/latest/schedulers/kubernetes_mcad.html
+
+## Non-Goals
+
+It is not a goal to provide an automated migration.
+Downstream projects and end-users will have to address the impacts manually to perform the migration to the new API groups.
+
+## How
+
+The progressive migration path is implemented in three phases as follows:
+
+### Phase 1: Add Dual API Groups Support
+
+The new API groups are introduced in MCAD, along with the existing ones.
+Functionally, the MCAD and InstaScale controllers are updated, so they can equally reconcile the old and new APIs.
+
+To implement that dual API groups support, it may be necessary to duplicate the MCAD API packages, as well as the generated client packages.
+Indeed, [controller-runtime](https://github.com/kubernetes-sigs/controller-runtime) APIs, which are used by InstaScale, are designed to use the API structs, and require the mapping between the API structs and the Group/Version/Kind(GVK)s to be an injection.
+Otherwise stated, controller-runtime mandates the scheme holding the mapping between the API structs and the GVKs to have a single GVK entry per API struct.
+
+This boils down to the following tasks:
+
+* Duplicate the API packages, with the new API groups, and mark the old ones as deprecated
```

Comment:
With the desire to make migration easier, is there a possibility to declare a current target MCAD release compatible with "all" downstream projects for the short term and not do duplication? 
Most downstream integrations are managed by the MCAD team, later individual integrators can work on the migration path.

Reply:
I agree we should lower duplication as much as possible.

I did not realize the TorchX MCAD scheduler is currently delivered from the CodeFlare fork of TorchX. That makes all the impacted components under our control, as you said.

I've added the _one-shot_ migration alternative in 5735b236e46708af4591aaa08c57a5b19eddbbcb, and given and example of how that could go.

Reply:
If the consensus is to go with the _on-shot_ migration alternative, I'll flip it with the _progressive_ migration proposal.

Reply:
If one-shot migration is simpler, I think we should bite the bullet and I vote for doing it. a stop-gap solution is to ask users to use let's say release 1.32 which has old CRDs.

Reply:
I agree, I'm a +1 for one-shot

Reply:
+1 on that also.

Reply:
Thanks all for the feedback, I've promoted the _one-shot_ migration as the proposed plan in 2c3eefe2cf8c7bff02923a7bfcba204db4e8bdbd.